persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stores/binary/JdbcBinaryStore.java

```diff
@@ -357,6 +309,101 @@ public void purge(Executor threadPool, PurgeListener task) {
+
+         for (Bucket bucket : existingBuckets.values()) {
+            if (newBuckets.contains(bucket)) {
+               prepareWriteStatement(insertBatch, bucket, tableManager.getInsertRowSql());
+               insertBatch.addBatch();
+            } else {
+               prepareWriteStatement(updateBatch, bucket, tableManager.getUpdateRowSql());
+               updateBatch.addBatch();
+            }
+         }
+         insertBatch.executeBatch();
+         updateBatch.executeBatch();
+      }
+   }
+
+   private Map<Integer, Bucket> getExistingBuckets(Connection connection, List<MarshalledEntry> modifiedEntries,
+                                                   List<Object> keysToDelete) throws SQLException {
+      List<Integer> bucketIds = Stream.concat(modifiedEntries.stream().map(MarshalledEntry::getKey), keysToDelete.stream())
+            .map(this::getBuckedId)
+            .distinct()
+            .collect(Collectors.toList());
+
+      String sql = tableManager.getSelectMultipleRowSql(bucketIds.size());
+      if (sql == null)
+         return new HashMap<>();
+
+      try (PreparedStatement ps = connection.prepareStatement(sql)) {
+         for (int i = 0; i < bucketIds.size(); i++) {
+            ps.setInt(i + 1, bucketIds.get(i));
+         }
+
+         try (ResultSet rs = ps.executeQuery()) {
+            Map<Integer, Bucket> existingBuckets = new HashMap<>();
+            while (rs.next()) {
+               Bucket bucket = loadBucket(rs);
+               existingBuckets.put(bucket.getBucketId(), bucket);
+            }
+            return existingBuckets;
+         }
+      }
+   }
+
+   // Returns a set of newly created Buckets
+   private Set<Bucket> updateAndCreateBuckets(List<MarshalledEntry> modifiedEntries, List<Object> keysToDelete,
+                                              Map<Integer, Bucket> existingBuckets) {
+      Set<Bucket> newBuckets = new HashSet<>();
+      for (MarshalledEntry entry : modifiedEntries) {
+         Integer bucketKey = getBuckedId(entry.getKey());
+         Object entryKey = entry.getKey();
+         InternalMetadata m = entry.getMetadata();
+         Bucket existingBucket = existingBuckets.get(bucketKey);
+
+         // How should we handle expired entries in a Tx?
+         if (m != null && m.isExpired(ctx.getTimeService().wallClockTime())) {
+            if (existingBucket != null) {
+               existingBucket.removeEntry(entryKey);
+            }
+            continue;
+         }
+
+         if (existingBucket == null) {
+            Bucket bucket = new Bucket(keyEquivalence);
+            bucket.setBucketId(bucketKey);
+            bucket.addEntry(entryKey, entry);
+            existingBuckets.put(bucketKey, bucket);
+            newBuckets.add(bucket);
+         } else {
+            existingBucket.addEntry(entryKey, entry);
+         }
+      }
+
+      // Remove keys from bucket
+      for (Object entryKey : keysToDelete) {
```

Comment:
What would happen if you have a remove and a put in that order?  Will the remove wipe out the modification?